crates/red_knot_test/src/assertion.rs

```diff
@@ -308,9 +341,169 @@ impl std::fmt::Display for ErrorAssertion<'_> {
+    fn skip_whitespace(&mut self) {
+        self.cursor.eat_while(char::is_whitespace);
+    }
+
+    /// Consume characters in the assertion comment until we find a character that satisfies `end_predicate`.
+    ///
+    /// Return a string slice representing the concatenation of all characters we consumed,
+    /// or `None` if we hit the end of the comment before finding a character
+    /// that satisfies `end_predicate`.
+    fn consume_until(&mut self, mut end_predicate: impl FnMut(char) -> bool) -> Option<&'a str> {
+        let start = self.offset().to_usize();
+
+        while !self.cursor.is_eof() {
+            if end_predicate(self.cursor.first()) {
+                return Some(&self.comment_source[start..self.offset().to_usize()]);
+            }
+            self.cursor.bump();
+        }
+
+        None
+    }
+
+    /// Attempt to parse the assertion comment into a [`ErrorAssertion`].
+    fn parse(mut self) -> Result<ErrorAssertion<'a>, ErrorAssertionParseError<'a>> {
+        let mut column = None;
+        let mut rule = None;
+
+        self.skip_whitespace();
+
+        while !self.cursor.is_eof() {
+            match self.cursor.first() {
+                // column number
+                '0'..='9' => {
+                    if column.is_some() {
+                        return Err(ErrorAssertionParseError::MultipleColumnNumbers);
+                    }
+                    if rule.is_some() {
+                        return Err(ErrorAssertionParseError::ColumnNumberAfterRuleCode);
+                    }
+                    // This will return `None` if it hits the end of the file before hitting any whitespace
+                    let Some(column_str) = self.consume_until(char::is_whitespace) else {
+                        return Err(ErrorAssertionParseError::NoRuleOrMessage);
+                    };
+                    column = OneIndexed::from_str(column_str)
+                        .map(Some)
+                        .map_err(|e| ErrorAssertionParseError::BadColumnNumber(column_str, e))?;
+                }
+
+                // rule code
+                '[' => {
+                    if rule.is_some() {
+                        return Err(ErrorAssertionParseError::MultipleRuleCodes);
+                    }
+                    self.cursor.bump();
+                    let Some(position) = memchr::memmem::find(self.cursor.as_bytes(), b"]") else {
+                        return Err(ErrorAssertionParseError::UnclosedRuleCode);
+                    };
+                    rule = Some(&self.cursor.as_str()[..position]);
+                    self.cursor.skip_bytes(position + b"]".len());
+                }
+
+                // message text
+                '"' => {
+                    self.cursor.bump();
+                    let rest = &self.comment_source
+                        [self.offset().to_usize()..self.comment_source.trim_end().len() - 1];
```

Comment:
```suggestion
                        [self.offset().to_usize()..self.comment_source.trim_end().len() - 1];
```

This could panic if the string doesn't end with a `"` but the last character is a multi bytes character. Why not `self.eath_while(|c| c != '"')`?

Reply:
> This could panic if the string doesn't end with a `"` but the last character is a multi bytes character.

good point, thanks!

> Why not `self.eath_while(|c| c != '"')`?

that doesn't work because some of our revealed types and error messages include `"` characters (e.g. `Literal["foo"]`). But I can do something else.